File: services/config/facade.ts
/**
 * Mossen 配置门面 (G1-4) — provider 链 + 公共 API.
 *
 * Provider 优先级 (按 Allen G-D1 决策):
 *   override (0) > env (1) > project (2) > user (3) > remote (4) > default (5)
 *
 * 远程 provider G1 阶段不实现 (G-D2 接口预留 default disabled);
 * G6 完成后 Mossen hosted 实现时再注册.
 */

import {
  PROVIDER_PRIORITY,
  type ConfigOverrideScope,
  type ConfigValueSource,
  type MossenConfigFacade,
  type MossenConfigProvider,
  type MossenConfigRefreshListener,
  type ProviderResult,
} from './types.js'
import { MOSSEN_BUILTIN_DEFAULTS } from './defaults.js'
import { resolveAliasedKey } from './aliasMap.js'
import {
  LocalDefaultProvider,
  ProjectSettingsProvider,
  UserSettingsProvider,
} from './providers/local.js'
import { EnvOverrideProvider } from './providers/envOverride.js'

/** Process-内 override (优先级最高) */
class RuntimeOverrideProvider implements MossenConfigProvider {
  readonly name: ConfigValueSource = 'override'

  readonly priority = PROVIDER_PRIORITY.override

  readonly enabled = true

  private readonly store = new Map<string, unknown>()

  get<T>(key: string): ProviderResult<T> {
    if (this.store.has(key)) {
      return { value: this.store.get(key) as T, source: 'override' }
    }
    return undefined
  }

  set<T>(key: string, value: T): void {
    this.store.set(key, value)
  }

  clear(key?: string): void {
    if (key === undefined) {
      this.store.clear()
    } else {
      this.store.delete(key)
    }
  }
}

const runtimeOverride = new RuntimeOverrideProvider()
const envOverride = new EnvOverrideProvider()
const projectSettings = new ProjectSettingsProvider()
const userSettings = new UserSettingsProvider()
const localDefault = new LocalDefaultProvider()

/**
 * Provider 链, 按 priority 升序排列.
 * G1 阶段 5 个 (无 remote); G6 后接入 remote 时插入到 user 之后.
 */
const PROVIDERS: readonly MossenConfigProvider[] = [
  runtimeOverride,
  envOverride,
  projectSettings,
  userSettings,
  localDefault,
]

/** 用于 set/clear 时按 scope 路由到具体 provider */
function pickWritableProvider(scope: ConfigOverrideScope): MossenConfigProvider {
  switch (scope) {
    case 'override':
      return runtimeOverride
    case 'project':
      return projectSettings
    case 'user':
      return userSettings
  }
}

const refreshListeners = new Set<MossenConfigRefreshListener>()

/** 通知所有 refresh listener (内部用; G6 后远程 provider 触发) */
export function notifyRefreshListeners(): void {
  for (const listener of refreshListeners) {
    try {
      const result = listener()
      if (result instanceof Promise) {
        result.catch(() => {
          // listener 自己 swallow; 不阻塞其他 listener
        })
      }
    } catch {
      // listener 抛错不阻塞其他 listener
    }
  }
}

/**
 * Resolve 给定 key, 顺序遍历 provider 链, 首个命中返回.
 * 全部 miss → 返回 default fallback (caller 传的 defaultValue).
 */
function resolve<T>(
  key: string,
  defaultValue: T,
): { value: T; source: ConfigValueSource; resolvedKey?: string } {
  // alias 解析: tengu_* → mossen.*
  const resolvedKey = resolveAliasedKey(key)
  const isAliased = resolvedKey !== key

  for (const provider of PROVIDERS) {
    if (!provider.enabled) continue
    const result = provider.get<T>(resolvedKey)
    if (result !== undefined) {
      return {
        value: result.value,
        source: result.source,
        ...(isAliased ? { resolvedKey } : {}),
      }
    }
  }
  return {
    value: defaultValue,
    source: 'default',
    ...(isAliased ? { resolvedKey } : {}),
  }
}

// ===== Facade public API =====

export function getMossenFeatureValue<T>(key: string, defaultValue: T): T {
  return resolve(key, defaultValue).value
}

export function getMossenDynamicConfig<T>(key: string, defaultValue: T): T {
  return resolve(key, defaultValue).value
}

export function checkMossenGate(key: string, defaultValue = false): boolean {
  const v = resolve<unknown>(key, defaultValue).value
  return Boolean(v)
}

export function onMossenConfigRefresh(
  listener: MossenConfigRefreshListener,
): () => void {
  refreshListeners.add(listener)
  return () => {
    refreshListeners.delete(listener)
  }
}

export function setMossenConfigOverride(
  key: string,
  value: unknown,
  scope: ConfigOverrideScope = 'override',
): void {
  const provider = pickWritableProvider(scope)
  provider.set?.(key, value)
  notifyRefreshListeners()
}

export function clearMossenConfigOverrides(
  scope: ConfigOverrideScope = 'override',
  key?: string,
): void {
  const provider = pickWritableProvider(scope)
  provider.clear?.(key)
  notifyRefreshListeners()
}

export function getAllMossenConfigValues(): Record<string, unknown> {
  // 遍历内置 default 表的所有 key, resolve 一遍
  const out: Record<string, unknown> = {}
  for (const key of Object.keys(MOSSEN_BUILTIN_DEFAULTS)) {
    out[key] = resolve(key, MOSSEN_BUILTIN_DEFAULTS[key]).value
  }
  return out
}

export function resolveMossenConfig<T>(
  key: string,
  defaultValue: T,
): { value: T; source: ConfigValueSource; resolvedKey?: string } {
  return resolve(key, defaultValue)
}

/** 仅供测试: reset 全部 in-memory 状态 */
export function _resetFacadeForTesting(): void {
  runtimeOverride.clear()
  refreshListeners.clear()
}

/** 自检: facade 实现了 MossenConfigFacade 接口 */
export const facade: MossenConfigFacade = {
  getMossenFeatureValue,
  getMossenDynamicConfig,
  checkMossenGate,
  onMossenConfigRefresh,
  setMossenConfigOverride,
  clearMossenConfigOverrides,
  getAllMossenConfigValues,
  resolveMossenConfig,
}

File: crates/mossen-agent/src/services/config/facade.rs
//! Configuration facade — provider chain + public API.
//!
//! Provider priority (override > env > project > user > default):
//!   override(0) > env(1) > project(2) > user(3) > default(5)

use once_cell::sync::Lazy;
use parking_lot::RwLock;
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Arc;

use super::alias_map::resolve_aliased_key;
use super::defaults::MOSSEN_BUILTIN_DEFAULTS;
use super::providers::env_override::EnvOverrideProvider;
use super::providers::local::{LocalDefaultProvider, ProjectSettingsProvider, UserSettingsProvider};
use super::types::{ConfigOverrideScope, ConfigValueSource, MossenConfigProvider, ProviderResult};

/// Process-internal runtime override provider (highest priority).
struct RuntimeOverrideProvider {
    store: RwLock<HashMap<String, Value>>,
}

impl RuntimeOverrideProvider {
    fn new() -> Self {
        Self {
            store: RwLock::new(HashMap::new()),
        }
    }
}

impl MossenConfigProvider for RuntimeOverrideProvider {
    fn name(&self) -> ConfigValueSource {
        ConfigValueSource::Override
    }

    fn priority(&self) -> u8 {
        ConfigValueSource::Override.priority()
    }

    fn enabled(&self) -> bool {
        true
    }

    fn get(&self, key: &str) -> Option<ProviderResult> {
        let store = self.store.read();
        store.get(key).map(|value| ProviderResult {
            value: value.clone(),
            source: ConfigValueSource::Override,
            resolved_key: None,
        })
    }

    fn set(&self, key: &str, value: Value) {
        let mut store = self.store.write();
        store.insert(key.to_string(), value);
    }

    fn clear(&self, key: Option<&str>) {
        let mut store = self.store.write();
        match key {
            None => store.clear(),
            Some(k) => {
                store.remove(k);
            }
        }
    }
}

struct FacadeInner {
    runtime_override: Arc<RuntimeOverrideProvider>,
    env_override: Arc<EnvOverrideProvider>,
    project_settings: Arc<ProjectSettingsProvider>,
    user_settings: Arc<UserSettingsProvider>,
    local_default: Arc<LocalDefaultProvider>,
    refresh_listeners: RwLock<Vec<Arc<dyn Fn() + Send + Sync>>>,
}

static FACADE: Lazy<FacadeInner> = Lazy::new(|| FacadeInner {
    runtime_override: Arc::new(RuntimeOverrideProvider::new()),
    env_override: Arc::new(EnvOverrideProvider),
    project_settings: Arc::new(ProjectSettingsProvider::new()),
    user_settings: Arc::new(UserSettingsProvider::new()),
    local_default: Arc::new(LocalDefaultProvider),
    refresh_listeners: RwLock::new(Vec::new()),
});

impl FacadeInner {
    fn providers(&self) -> Vec<&dyn MossenConfigProvider> {
        vec![
            self.runtime_override.as_ref() as &dyn MossenConfigProvider,
            self.env_override.as_ref() as &dyn MossenConfigProvider,
            self.project_settings.as_ref() as &dyn MossenConfigProvider,
            self.user_settings.as_ref() as &dyn MossenConfigProvider,
            self.local_default.as_ref() as &dyn MossenConfigProvider,
        ]
    }

    fn pick_writable_provider(&self, scope: ConfigOverrideScope) -> &dyn MossenConfigProvider {
        match scope {
            ConfigOverrideScope::Override => self.runtime_override.as_ref(),
            ConfigOverrideScope::Project => self.project_settings.as_ref(),
            ConfigOverrideScope::User => self.user_settings.as_ref(),
        }
    }

    fn notify_refresh_listeners(&self) {
        let listeners = self.refresh_listeners.read();
        for listener in listeners.iter() {
            listener();
        }
    }
}

/// Resolve a key through the provider chain.
fn resolve(key: &str, default_value: Value) -> ResolveResult {
    let resolved_key = resolve_aliased_key(key);
    let is_aliased = resolved_key != key;

    for provider in FACADE.providers() {
        if !provider.enabled() {
            continue;
        }
        if let Some(result) = provider.get(resolved_key) {
            return ResolveResult {
                value: result.value,
                source: result.source,
                resolved_key: if is_aliased {
                    Some(resolved_key.to_string())
                } else {
                    None
                },
            };
        }
    }
    ResolveResult {
        value: default_value,
        source: ConfigValueSource::Default,
        resolved_key: if is_aliased {
            Some(resolved_key.to_string())
        } else {
            None
        },
    }
}

/// Result of resolving a configuration key.
#[derive(Debug, Clone)]
pub struct ResolveResult {
    pub value: Value,
    pub source: ConfigValueSource,
    pub resolved_key: Option<String>,
}

// ===== Public facade API =====

/// Get a feature value by key.
pub fn get_mossen_feature_value(key: &str, default_value: Value) -> Value {
    resolve(key, default_value).value
}

/// Get a dynamic config by key.
pub fn get_mossen_dynamic_config(key: &str, default_value: Value) -> Value {
    resolve(key, default_value).value
}

/// Check a gate (boolean).
pub fn check_mossen_gate(key: &str, default_value: bool) -> bool {
    let v = resolve(key, Value::Bool(default_value)).value;
    match v {
        Value::Bool(b) => b,
        _ => default_value,
    }
}

/// Subscribe to config refresh events. Returns unsubscribe closure.
pub fn on_mossen_config_refresh(listener: Arc<dyn Fn() + Send + Sync>) -> impl FnOnce() {
    let listener_clone = Arc::clone(&listener);
    FACADE.refresh_listeners.write().push(listener_clone);
    let listener_ptr = Arc::as_ptr(&listener) as *const () as usize;
    move || {
        let mut listeners = FACADE.refresh_listeners.write();
        listeners.retain(|l| Arc::as_ptr(l) as *const () as usize != listener_ptr);
    }
}

/// Set a config override.
pub fn set_mossen_config_override(key: &str, value: Value, scope: ConfigOverrideScope) {
    let provider = FACADE.pick_writable_provider(scope);
    provider.set(key, value);
    FACADE.notify_refresh_listeners();
}

/// Clear config overrides.
pub fn clear_mossen_config_overrides(scope: ConfigOverrideScope, key: Option<&str>) {
    let provider = FACADE.pick_writable_provider(scope);
    provider.clear(key);
    FACADE.notify_refresh_listeners();
}

/// Get all resolved config values (built-in defaults only).
pub fn get_all_mossen_config_values() -> HashMap<String, Value> {
    let mut out = HashMap::new();
    for (key, default_val) in MOSSEN_BUILTIN_DEFAULTS.iter() {
        let resolved = resolve(key, default_val.clone());
        out.insert(key.to_string(), resolved.value);
    }
    out
}

/// Resolve a config key with full metadata.
pub fn resolve_mossen_config(key: &str, default_value: Value) -> ResolveResult {
    resolve(key, default_value)
}

/// Notify all refresh listeners (internal use).
pub fn notify_refresh_listeners() {
    FACADE.notify_refresh_listeners();
}

/// Reset facade for testing.
pub fn reset_facade_for_testing() {
    FACADE.runtime_override.clear(None);
    FACADE.refresh_listeners.write().clear();
}
